### ISSUE TITLE
Test sort stability

### ISSUE THREAD

**Issue by treeowl** [haskell/containers#509]
`sort`, `sortBy`, and `sortOn` are all supposed to be stable sorts, but we only test stability for `sortBy`.

**Comment by treeowl** [comment]
I'm pretty sure there is a type defined somewhere in the set/map tests with a sloppy `Ord` instance specifically designed for these sorts of tests.

**Comment by jwaldmann** [comment]
bit late to the party (just saw the note in HWN), but .. I'd do it like this, where
* "hiding" is a stand-alone concept (not fused with "pairing")
* we check for monotonicity of the output only (assuming that it's checked elsewhere that the function returns the correct multiset)

```
{-# language RankNTypes #-}

import Test.LeanCheck

import qualified Data.Sequence as S
import Data.Foldable (toList)

newtype Hide t = Hide { reveal :: t } deriving Show

instance Eq (Hide t) where _ == _ = True
instance Ord (Hide t) where compare _ _  = EQ

sort_is_stable :: (forall a . Ord a => S.Seq a -> S.Seq a) -> [Int ] -> Bool
sort_is_stable  sort xs =
  let input = zip xs $ map Hide [ 0 :: Int .. ]
      output = toList $ sort $ S.fromList input
  in  monotone $ map ( \(x,y) -> (x, reveal y) ) output

monotone xs = and $ zipWith (<=) xs $ tail xs

main = do
  checkFor 10000 $ sort_is_stable S.sort
  checkFor 10000 $ sort_is_stable S.unstableSort
```

**Comment by treeowl** [comment]
That's too coarse.

**Comment by jwaldmann** [comment]
Out of curiosity - what is the exact definition of "stability" that these sort functions realize?

This definition could be put in the haddocks, so programmers know exactly what to expect. 

Here is one example of a formal definition of "stability w.r.t. a key function" http://dx.doi.org/10.1007/s10817-012-9260-7 (Def. 4)

**Comment by treeowl** [comment]
Fixed in #514

**Comment by treeowl** [comment]
@jwaldmann, I'm not sure we want to go into so much detail about a relatively well-known idea. Can you suggest something brief and understandable?
